zenoh-flow/src/io/input.rs

```diff
@@ -0,0 +1,190 @@
+    /// associated, otherwise `None` is returned.
+    ///
+    /// ## Performance
+    ///
+    /// With a typed [`Input<T>`](`Input`), Zenoh-Flow will perform operations on the underlying
+    /// data: if the data are received serialized then Zenoh-Flow will deserialize them, if they are
+    /// received "typed" then Zenoh-Flow will check that the type matches what is expected.
+    ///
+    /// If the underlying data is not relevant then these additional operations can be avoided by
+    /// calling `take_raw` and using an [`InputRaw`](`InputRaw`) instead.
+    pub fn take<T: ZFData>(&mut self, port_id: impl AsRef<str>) -> Option<Input<T>> {
+        self.hmap.remove(port_id.as_ref()).map(|receivers| Input {
+            _phantom: PhantomData,
+            input_raw: InputRaw {
+                port_id: port_id.as_ref().into(),
+                receivers,
+            },
+        })
+    }
+
+    /// Returns the [`InputRaw`](`InputRaw`) associated to the provided `port_id`, if one is
+    /// associated, otherwise `None` is returned.
+    ///
+    /// ## Convenience
+    ///
+    /// With an [`InputRaw`](`InputRaw`), Zenoh-Flow will not manipulate the underlying data, for
+    /// instance trying to deserialize them to a certain `T`.
+    ///
+    /// It is thus up to the user to call `try_get::<T>()` on the [`Payload`](`Payload`) and handle
+    /// the error that could surface.
+    ///
+    /// If all the data must be "converted" to the same type `T` then calling `take::<T>()` and
+    /// using a typed [`Input`](`Input`) will be more convenient and as efficient.
+    pub fn take_raw(&mut self, port_id: impl AsRef<str>) -> Option<InputRaw> {
+        self.hmap
+            .remove(port_id.as_ref())
+            .map(|receivers| InputRaw {
+                port_id: port_id.as_ref().into(),
+                receivers,
+            })
+    }
+}
+
+/// An [`InputRaw`](`InputRaw`) exposes the [`LinkMessage`](`LinkMessage`) it receives.
+///
+/// It's primary purpose is to ensure "optimal" performance. This can be useful to implement
+/// behaviour where actual access to the underlying data is irrelevant.
+#[derive(Clone, Debug)]
+pub struct InputRaw {
+    pub(crate) port_id: PortId,
+    pub(crate) receivers: Vec<flume::Receiver<LinkMessage>>,
+}
+
+impl InputRaw {
+    pub fn port_id(&self) -> &PortId {
+        &self.port_id
+    }
+
+    /// Returns the number of channels associated with this Input.
+    pub fn channels_count(&self) -> usize {
+        self.receivers.len()
+    }
+
+    /// Returns the first [`LinkMessage`](`LinkMessage`) that was received, *asynchronously*, on any
+    /// of the channels associated with this Input.
+    ///
+    /// If several [`LinkMessage`](`LinkMessage`) are received at the same time, one is *randomly*
+    /// selected.
+    ///
+    /// ## Error
+    ///
+    /// If an error occurs on one of the channels, this error is returned.
+    pub async fn recv_async(&self) -> Result<LinkMessage> {
```

Comment:
Are we going to add a `try_recv()` as briefly discussed?

Reply:
I did not want to "mix" too many changes in the same PR. As we want to remove the, potentially harmful, synchronous `recv` I saw no point of updating it to the changes and so removed it now. But this leaves things in an "half-done" state, so I'll add the `try_{recv|send}` and rename the methods `{recv|send}_async` to `{recv|send}`. :)